cpp/src/arrow/compute/kernels/aggregate_mode.cc

```diff
@@ -378,6 +395,21 @@ VectorKernel NewModeKernel(const std::shared_ptr<DataType>& in_type) {
   return kernel;
 }
 
+Result<ValueDescr> ModeType(KernelContext*, const std::vector<ValueDescr>& descrs) {
+  return ValueDescr::Array(
+      struct_({field(kModeFieldName, descrs[0].type), field(kCountFieldName, int64())}));
+}
+
+VectorKernel NewModeKernel(Type::type type_id, ArrayKernelExec exec) {
```

Comment:
Nit: perhaps define a single `NewModeKernel(InputType, OutputType, ArrayKernelExec)`?

Reply:
Updated the registration here and the docs below.